choosenim does not work on Linux with musl

As in the title, trying to run `curl https://nim-lang.org/choosenim/init.sh -sSf | sh` on a musl based Linux distro results in:

```
choosenim-init: Downloading choosenim-0.6.0_linux_amd64
sh: /tmp/choosenim-0.6.0_linux_amd64: not found
```

This is because choosenim is linked against glibc:

```
thermi:~$ ldd /tmp/choosenim-0.6.0_linux_amd64 
        /lib64/ld-linux-x86-64.so.2 (0x7fc250e12000)
        librt.so.1 => /lib64/ld-linux-x86-64.so.2 (0x7fc250e12000)
        libdl.so.2 => /lib64/ld-linux-x86-64.so.2 (0x7fc250e12000)
        libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x7fc250e12000)
Error relocating /tmp/choosenim-0.6.0_linux_amd64: __fdelt_chk: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: __poll_chk: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: __strdup: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: __memcpy_chk: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: gnu_dev_major: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: __longjmp_chk: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: gnu_dev_minor: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: __fprintf_chk: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: gnu_dev_makedev: symbol not found
Error relocating /tmp/choosenim-0.6.0_linux_amd64: __sprintf_chk: symbol not found
```

So, musl is out because libarchive cannot work with it:

```
/home/dom/.cache/nim/nimterop/nimarchive/libarchive/LibArchive/lib/libarchive.a(archive_entry.o): In function `archive_entry_devmajor':
        ... /workspace/srcdir/libarchive/libarchive/archive_entry.c:344: undefined reference to `gnu_dev_major'

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,12 +6,18 @@ os:
 language: c
 
 env:
-  - BRANCH=1.2.4
+  - BRANCH=1.2.6
 
 cache:
   directories:
     - "$HOME/.choosenim"
     - "$TRAVIS_BUILD_DIR/git"
+
+addons:
+  apt:
+    update: true
+    packages:
+      - musl-tools
 
 install:
   # Use common Travis script maintained as a gist

--- a/choosenim.nimble
+++ b/choosenim.nimble
@@ -13,7 +13,8 @@ skipExt = @["nim"]
 
 # Dependencies
 
-requires "nim >= 1.2.4", "nimble#14a6946", "nimarchive >= 0.5.2"
+requires "nim >= 1.2.6", "nimble#26167cd"
+requires "nimterop >= 0.6.12", "nimarchive >= 0.5.3"
 requires "libcurl >= 1.0.0"
 requires "analytics >= 0.2.0"
 requires "osinfo >= 0.3.0"

--- a/src/choosenim.nim
+++ b/src/choosenim.nim
@@ -7,7 +7,7 @@ import nimblepkg/common as nimbleCommon
 from nimblepkg/packageinfo import getNameVersion
 
 import choosenimpkg/[download, builder, switcher, common, cliparams, versions]
-import choosenimpkg/[utils, channel, telemetry]
+import choosenimpkg/[utils, channel, ssl, telemetry]
 
 when defined(windows):
   import choosenimpkg/env

--- a/src/choosenim.nims
+++ b/src/choosenim.nims
@@ -1,4 +1,12 @@
-when defined(macosx):
-  --define:curl
-else:
-  --define:ssl
+when findExe("musl-gcc").len != 0:
+  # Use musl-gcc when available
+  putEnv("CC", "musl-gcc")
+  switch("gcc.exe", "musl-gcc")
+  switch("gcc.linkerexe", "musl-gcc")
+
+# Statically linking everything
+when not defined(OSX):
+  switch("passL", "-static")
+switch("define", "ssl")
+switch("dynlibOverride", "ssl")
+switch("dynlibOverride", "crypto")

--- a/src/choosenimpkg/ssl.nim
+++ b/src/choosenimpkg/ssl.nim
@@ -1,0 +1,20 @@
+import strutils
+
+import nimterop/[build, cimport]
+
+# Download openssl from JuliaBinaryWrappers
+setDefines(@[
+  "cryptoJBB", "cryptoStatic"
+])
+
+getHeader(
+  "crypto.h",
+  jbburi = "openssl",
+  outdir = getProjectCacheDir("nimopenssl")
+)
+
+const
+  sslLPath = cryptoLPath.replace("crypto", "ssl")
+
+# Link static binaries
+{.passL: sslLPath & " " & cryptoLPath.}

